x509util/sctcheck/sctcheck.go

```diff
@@ -0,0 +1,242 @@
+	if len(chain) < 2 {
+		glog.Info("No issuer in chain; attempting online retrieval")
+		var err error
+		issuer, err = x509util.GetIssuer(leaf, hc)
+		if err != nil {
+			glog.Errorf("Failed to get issuer online: %v", err)
+		}
+	} else {
+		issuer = chain[1]
+	}
+
+	// Build a Merkle leaf that corresponds to the embedded SCTs.  We can use the same
+	// leaf for all of the SCTs, as long as the timestamp field gets updated.
+	merkleLeaf, err := ct.MerkleTreeLeafForEmbeddedSCT([]*x509.Certificate{leaf, issuer}, 0)
+	if err != nil {
+		glog.Errorf("Failed to build Merkle leaf: %v", err)
+		return 0, len(leaf.SCTList.SCTList)
+	}
+
+	var valid, invalid int
+	for i, sctData := range leaf.SCTList.SCTList {
+		subject := fmt.Sprintf("embedded SCT[%d]", i)
+		if checkSCT(ctx, lf, subject, merkleLeaf, sctData.Val, ll, hc) {
+			valid++
+		} else {
+			invalid++
+		}
+	}
+	return valid, invalid
+}
+
+func getAndCheckSiteChain(ctx context.Context, lf logInfoFactory, target string, ll *loglist.LogList, hc *http.Client) ([]*x509.Certificate, int, int, error) {
+	u, err := url.Parse(target)
+	if err != nil {
+		return nil, 0, 0, fmt.Errorf("failed to parse URL: %v", err)
+	}
+	if u.Scheme != "https" {
+		return nil, 0, 0, errors.New("non-https URL provided")
+	}
+	host := u.Host
+	if !strings.Contains(host, ":") {
+		host += ":443"
+	}
+
+	glog.Infof("Retrieve certificate chain from TLS connection to %q", host)
+	dialer := net.Dialer{Timeout: hc.Timeout}
+	conn, err := cryptotls.DialWithDialer(&dialer, "tcp", host, &cryptotls.Config{InsecureSkipVerify: true})
+	if err != nil {
+		return nil, 0, 0, fmt.Errorf("failed to dial %q: %v", host, err)
+	}
+	defer conn.Close()
+
+	goChain := conn.ConnectionState().PeerCertificates
+	glog.Infof("Found chain of length %d", len(goChain))
+
+	// Convert base crypto/x509.Certificates to our forked x509.Certificate type.
+	chain := make([]*x509.Certificate, len(goChain))
+	for i, goCert := range goChain {
+		cert, err := x509.ParseCertificate(goCert.Raw)
+		if err != nil {
+			return nil, 0, 0, fmt.Errorf("failed to convert Go Certificate [%d]: %v", i, err)
+		}
+		chain[i] = cert
+	}
+
+	// Check externally-provided SCTs.
+	var valid, invalid int
+	scts := conn.ConnectionState().SignedCertificateTimestamps
+	if len(scts) > 0 {
+		merkleLeaf, err := ct.MerkleTreeLeafFromChain(chain, ct.X509LogEntryType, 0 /* timestamp added later */)
+		if err != nil {
+			glog.Errorf("Failed to build Merkle tree leaf: %v", err)
+			return chain, 0, len(scts), nil
+		}
+		for i, sctData := range scts {
+			subject := fmt.Sprintf("external SCT[%d]", i)
+			if checkSCT(ctx, lf, subject, merkleLeaf, sctData, ll, hc) {
+				valid++
+			} else {
+				invalid++
+			}
+
+		}
+	}
+
+	return chain, valid, invalid, nil
+}
+
+func checkSCT(ctx context.Context, liFactory logInfoFactory, subject string, merkleLeaf *ct.MerkleTreeLeaf, sctData []byte, ll *loglist.LogList, hc *http.Client) bool {
+	var sct ct.SignedCertificateTimestamp
```

Comment:
Could use the new x509util.ExtractSCT function? (if you make checkSCT take a SerializedSCT instead of []byte)

Reply:
Done.